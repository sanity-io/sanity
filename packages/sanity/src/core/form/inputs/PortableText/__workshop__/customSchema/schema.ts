import {Rule} from '@sanity/types'
import {createSchema} from '../../../../../schema'

interface PTBlockValue {
  _type: 'block'
  _key: string
  children?: {_type: string; _key: string; text?: string}[]
}

const imageType = {
  type: 'image',
  name: 'blockImage',
}

const linkType = {
  type: 'object',
  name: 'link',
  fields: [
    {
      type: 'string',
      name: 'href',
    },
  ],
}

const someAnnotationType = {
  type: 'object',
  name: 'someAnnotation',
  options: {
    modal: {
      size: 'medium',
    },
  },

  fields: [{type: 'string', name: 'title'}],
}

const someObject = {
  type: 'object',
  name: 'someObject',
  options: {
    // modal: {type: 'dialog', width: 'auto'},
    modal: {
      type: 'dialog',
      size: 'full',
    },
  },

  fields: [{type: 'string', name: 'title'}],
}

function extractTextFromBlocks(blocks: PTBlockValue[]) {
  if (!blocks) {
    return ''
  }
  return blocks
    .filter((val) => val._type === 'block')
    .map((block) => {
      return (block.children || [])
        .filter((child) => child._type === 'span')
        .map((span) => span.text)
        .join('')
    })
    .join('')
}

export const blockType = {
  type: 'block',
  name: 'myBlockType',
  styles: [
    {title: 'Paragraph', value: 'normal'},
    {title: 'Heading 1', value: 'h1'},
    {title: 'Heading 2', value: 'h2'},
    {title: 'Heading 3', value: 'h3'},
    {title: 'Heading 4', value: 'h4'},
    {title: 'Heading 5', value: 'h5'},
    {title: 'Heading 6', value: 'h6'},
    {title: 'Quotation', value: 'blockquote'},
  ],

  marks: {
    annotations: [someAnnotationType, linkType],
  },

  of: [someObject, imageType],
  validation: (rule: Rule) =>
    rule
      .custom((block: PTBlockValue): any => {
        const length = extractTextFromBlocks([block]).length
        return length < 10 ? 'Please write a longer paragraph.' : null
      })
      .error(),
}

const ptType = {
  type: 'array',
  name: 'body',
  of: [blockType, someObject, imageType],
}

export const schema = createSchema({
  name: 'default',
  types: [imageType, someObject, ptType],
})

export const portableTextType = schema.get('body')
