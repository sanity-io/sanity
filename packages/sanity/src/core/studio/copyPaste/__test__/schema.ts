import {Schema as SchemaBuilder} from '@sanity/schema'
import {defineType, type Schema} from '@sanity/types'

const Icon = () => null

const linkType = defineType({
  type: 'object',
  name: 'link',
  fields: [
    {
      type: 'string',
      name: 'href',
      validation: (Rule) => Rule.uri({scheme: ['http', 'https']}),
    },
  ],
  validation: (Rule) => Rule.required(),
})

const myStringType = defineType({
  type: 'object',
  name: 'test',
  fields: [{type: 'string', name: 'mystring', validation: (Rule) => Rule.required()}],
})

export const schema = SchemaBuilder.compile({
  name: 'default',
  types: [
    linkType,
    myStringType,
    {
      name: 'customNamedBlock',
      type: 'block',
      title: 'A named custom block',
      marks: {
        annotations: [linkType, myStringType],
      },
      of: [
        {type: 'image'},
        {
          type: 'object',
          name: 'test',
          fields: [myStringType],
        },
        {
          type: 'reference',
          name: 'strongAuthorRef',
          title: 'A strong author ref',
          to: {type: 'author'},
        },
      ],
    },
    {
      name: 'author',
      title: 'Author',
      type: 'document',
      icon: Icon,
      fields: [
        {
          name: 'name',
          type: 'string',
        },
        {
          name: 'role',
          type: 'string',
        },
        {
          name: 'bio',
          type: 'array',
          of: [{type: 'customNamedBlock'}],
        },
        {
          name: 'bestFriend',
          type: 'reference',
          to: [{type: 'author'}],
        },
      ],
      initialValue: () => ({
        role: 'Developer',
      }),
    },
    {
      name: 'address',
      title: 'Address',
      type: 'object',
      fields: [
        {
          name: 'street',
          type: 'string',
          initialValue: 'one old street',
        },
        {
          name: 'streetNo',
          type: 'string',
          initialValue: '123',
        },
      ],
    },
    {
      name: 'contact',
      title: 'Contact',
      type: 'object',
      fields: [
        {
          name: 'email',
          type: 'string',
        },
        {
          name: 'phone',
          type: 'string',
        },
      ],
    },
    {
      name: 'person',
      title: 'Person',
      type: 'document',
      icon: Icon,
      fields: [
        {
          name: 'address',
          type: 'address',
        },
        {
          name: 'contact',
          type: 'contact',
        },
      ],
    },

    {
      name: 'post',
      title: 'Post',
      type: 'document',
      icon: Icon,
      fields: [
        {
          name: 'title',
          type: 'string',
        },
        {
          name: 'body',
          type: 'array',
          of: [{type: 'customNamedBlock'}],
        },
        {
          name: 'author',
          type: 'reference',
          to: [{type: 'author'}],
        },
      ],
    },

    {
      name: 'captionedImage',
      type: 'object',
      fields: [
        {
          // This doesn't have a default value, so shouldn't be present,
          // not even with a `_type` stub
          name: 'asset',
          type: 'reference',
          to: [{type: 'sanity.imageAsset'}],
        },
        {
          name: 'caption',
          type: 'string',
        },
      ],
      initialValue: {caption: 'Default caption!'},
    },

    {
      name: 'recursiveObject',
      type: 'object',
      fields: [
        {
          name: 'name',
          type: 'string',
        },
        {
          name: 'child',
          type: 'recursiveObject',
        },
      ],
      initialValue: {
        name: '∞ recursion is ∞',
      },
    },

    {
      name: 'developer',
      type: 'document',
      initialValue: () => ({
        name: 'A default name!',

        // Should clear the default value below (but ideally not actually be part
        // of the value, eg no `undefined` in the resolved value)
        numberOfCats: undefined,
      }),
      fields: [
        {
          name: 'name',
          type: 'string',
        },
        {
          name: 'hasPet',
          type: 'boolean',
          initialValue: false,
        },
        {
          name: 'age',
          type: 'number',
          initialValue: 30,
        },
        {
          name: 'numberOfCats',
          type: 'number',
          initialValue: 3,
        },
        {
          name: 'heroImage',
          type: 'captionedImage',
        },
        {
          name: 'awards',
          type: 'array',
          of: [{type: 'string'}],
          initialValue: () => ['TypeScript Wizard of the Year'],
        },
        {
          name: 'tasks',
          type: 'array',
          of: [
            {
              name: 'task',
              type: 'object',
              fields: [
                {name: 'description', type: 'string'},
                {name: 'isDone', type: 'boolean', initialValue: false},
              ],
            },
          ],
          initialValue: () => [
            {
              _type: 'task',
              description: 'Mark as done',
              isDone: false,
            },
          ],
        },
        {
          name: 'recursive',
          type: 'recursiveObject',
          // Initial value set on it's actual type
        },
      ],
    },
  ],
}) as Schema
