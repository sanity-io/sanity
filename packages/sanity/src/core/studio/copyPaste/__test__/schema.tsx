import {defineArrayMember, defineField, defineType, type Schema} from '@sanity/types'

import {createSchema} from '../../../schema'

const linkType = defineType({
  type: 'object',
  name: 'link',
  fields: [
    {
      type: 'string',
      name: 'href',
      validation: (Rule) => Rule.uri({scheme: ['http', 'https']}),
    },
  ],
  validation: (Rule) => Rule.required(),
})

const myStringObjectType = defineType({
  type: 'object',
  name: 'myStringObject',
  fields: [{type: 'string', name: 'myString', validation: (Rule) => Rule.required()}],
})

const nestedObjectType = defineType({
  type: 'object',
  name: 'nestedObject',
  fields: [
    {
      name: 'title',
      type: 'string',
    },
    {
      type: 'array',
      name: 'objectList',
      of: [{type: 'nestedObject'}],
    },
    {
      type: 'object',
      name: 'recursiveTest',
      fields: [
        {
          name: 'recursive',
          type: 'nestedObject',
        },
      ],
    },
  ],
})

export const schema = createSchema({
  name: 'default',
  types: [
    linkType,
    myStringObjectType,
    nestedObjectType,
    {
      name: 'customNamedBlock',
      type: 'block',
      title: 'A named custom block',
      marks: {
        annotations: [linkType, myStringObjectType],
      },
      of: [
        {
          type: 'object',
          name: 'test',
          fields: [myStringObjectType],
        },
        {
          type: 'reference',
          name: 'strongAuthorRef',
          title: 'A strong author ref',
          to: {type: 'author'},
        },
      ],
    },
    {
      name: 'author',
      title: 'Author',
      type: 'document',
      fields: [
        {
          name: 'name',
          type: 'string',
        },
        {
          name: 'uniqueStringNotInOtherDocument',
          type: 'string',
        },
        {
          name: 'isVerified',
          type: 'boolean',
        },
        {
          name: 'born',
          type: 'number',
        },
        {
          name: 'favoriteNumbers',
          type: 'array',
          of: [{type: 'number'}],
        },
        {
          name: 'testNumberWithListObjects',
          title: 'Test Number - List Objects',
          type: 'number',
          options: {
            list: [
              {value: 1, title: 'One'},
              {value: 2, title: 'Two'},
            ],
          },
        },
        {type: 'image', name: 'profileImage'},
        {
          type: 'object',
          name: 'socialLinks',
          fields: [
            {type: 'string', name: 'twitter'},
            {type: 'string', name: 'linkedin'},
          ],
        },
        {
          name: 'nestedTest',
          type: 'nestedObject',
        },
        {
          name: 'bio',
          type: 'array',
          of: [
            {type: 'customNamedBlock'},
            {type: 'myStringObject'},
            {type: 'reference', to: [{type: 'author'}]},
          ],
        },
        {
          name: 'friends',
          type: 'array',
          of: [{type: 'reference', to: [{type: 'author'}]}],
        },
        {
          name: 'bestFriend',
          type: 'reference',
          weak: true,
          to: [{type: 'author'}],
        },
      ],
    },
    {
      name: 'editor',
      title: 'Editor',
      type: 'document',
      fields: [
        {
          name: 'name',
          type: 'string',
        },
        {
          name: 'isVerified',
          type: 'boolean',
        },
        {
          name: 'born',
          type: 'number',
        },
        {type: 'image', name: 'profileImage'},
        {type: 'image', name: 'profileImagePNG', options: {accept: 'image/png'}},
        {type: 'image', name: 'profileImageJpeg', options: {accept: 'image/jpeg'}},
        {type: 'file', name: 'profileCV'},
        {
          name: 'bio',
          type: 'array',
          of: [{type: 'customNamedBlock'}],
        },
        {
          name: 'favoriteNumbers',
          type: 'array',
          of: [{type: 'number'}],
        },
        {
          name: 'favoriteStrings',
          type: 'array',
          of: [{type: 'string'}],
        },
        {
          name: 'favoriteBooleans',
          type: 'array',
          of: [{type: 'boolean'}],
        },
        {
          name: 'testNumberWithListObjects',
          title: 'Test Number - List Objects',
          type: 'number',
          options: {
            list: [
              {value: 1, title: 'One'},
              {value: 2, title: 'Two'},
            ],
          },
        },
        {
          name: 'nestedTest',
          type: 'nestedObject',
        },
        {
          name: 'profile',
          type: 'object',
          fields: [
            {type: 'string', name: 'email'},
            {type: 'boolean', name: 'isFavorite'},
            {type: 'image', name: 'avatar'},
            {
              type: 'object',
              name: 'social',
              fields: [
                {type: 'string', name: 'twitter'},
                {type: 'string', name: 'linkedin'},
              ],
            },
          ],
        },
        {
          name: 'friends',
          type: 'array',
          of: [{type: 'reference', to: [{type: 'editor'}, {type: 'author'}]}],
        },
        {
          name: 'bestAuthorFriend',
          type: 'reference',
          weak: false,
          to: [{type: 'author'}],
        },
        {
          name: 'relatedEditor',
          type: 'reference',
          weak: false,
          to: [{type: 'editor'}],
        },
        defineField({
          name: 'arrayWithAnonymousObject',
          title: 'Array with anonymous objects',
          description: 'This array contains objects of type as defined inline',
          type: 'array',
          of: [
            {
              type: 'object',
              title: 'Something',
              fields: [
                {name: 'first', type: 'string', title: 'First string'},
                {name: 'second', type: 'string', title: 'Second string'},
              ],
            },
          ],
        }),
        defineField({
          name: 'arrayOfPredefinedOptions',
          title: 'Array of predefined options',
          type: 'array',
          of: [
            {
              type: 'object',
              name: 'color',
              fields: [
                {
                  name: 'title',
                  type: 'string',
                },
                {
                  name: 'name',
                  type: 'string',
                },
              ],
            },
          ],
          options: {
            list: [
              {_type: 'color', title: 'Red', name: 'red'},
              {_type: 'color', title: 'Green', name: 'green', _key: 'green'},
              {_type: 'color', title: 'Blue', name: 'blue', _key: 'blue'},
              {_type: 'color', title: 'Black', name: 'black', _key: 'black'},
            ],
          },
        }),
        {
          name: 'arrayOfMultipleTypes',
          title: 'Array of multiple types',
          type: 'array',
          of: [
            {
              type: 'image',
            },
            {
              type: 'object',
              name: 'color',
              title: 'Color with a long title',
              fields: [
                {
                  name: 'title',
                  type: 'string',
                },
                {
                  name: 'name',
                  type: 'string',
                },
              ],
            },
          ],
        },
        defineField({
          name: 'arrayOfMultipleNestedTypes',
          title: 'Array of multiple nested types',
          type: 'array',
          of: [
            defineArrayMember({
              type: 'image',
            }),
            defineArrayMember({
              type: 'object',
              name: 'color',
              title: 'Color with a long title',
              fields: [
                {
                  name: 'title',
                  type: 'string',
                },
                {
                  name: 'name',
                  type: 'string',
                },
                defineField({
                  name: 'nestedArray',
                  title: 'Nested array',
                  type: 'array',
                  of: [
                    defineArrayMember({
                      type: 'object',
                      name: 'color',
                      title: 'Nested color with a long title',
                      fields: [
                        {
                          name: 'title',
                          type: 'string',
                        },
                        {
                          name: 'name',
                          type: 'string',
                        },
                      ],
                    }),
                  ],
                }),
              ],
            }),
          ],
        }),
      ],
    },
    {
      name: 'post',
      title: 'Post',
      type: 'document',
      fields: [
        {
          name: 'title',
          type: 'string',
        },
        {
          name: 'body',
          type: 'array',
          of: [{type: 'customNamedBlock'}],
        },
        {
          name: 'author',
          type: 'reference',
          to: [{type: 'author'}],
        },
      ],
    },
  ],
}) as Schema
