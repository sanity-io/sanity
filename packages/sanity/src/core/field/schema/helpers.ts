import {ArraySchemaType, isTypedObject, SchemaType} from '@sanity/types'
import {Diff} from '../types'

export function resolveTypeName(value: unknown): string {
  return isTypedObject(value) ? value._type : resolveJSType(value)
}

export function getArrayDiffItemType(
  diff: Diff,
  schemaType: ArraySchemaType,
): {fromType?: SchemaType; toType?: SchemaType} {
  if (diff.action === 'added') {
    return {
      toType: resolveArrayMemberType(schemaType, diff.toValue),
    }
  }

  if (diff.action === 'changed') {
    return {
      fromType: resolveArrayMemberType(schemaType, diff.fromValue),
      toType: resolveArrayMemberType(schemaType, diff.toValue),
    }
  }

  if (diff.action === 'removed') {
    return {
      fromType: resolveArrayMemberType(schemaType, diff.fromValue),
    }
  }

  // unchanged
  return {
    toType: resolveArrayMemberType(schemaType, diff.toValue),
  }
}

function resolveArrayMemberType(
  schemaType: ArraySchemaType,
  value: unknown,
): SchemaType | undefined {
  const typeName = resolveTypeName(value)
  const declared = schemaType.of.find((candidate) => candidate.name === typeName)
  if (declared) {
    return declared as any
  }

  return (schemaType.of.length === 1 ? schemaType.of[0] : undefined) as any
}

function resolveJSType(val: unknown) {
  if (Array.isArray(val)) {
    return 'array'
  }

  if (val === null) {
    return 'null'
  }

  return typeof val
}
