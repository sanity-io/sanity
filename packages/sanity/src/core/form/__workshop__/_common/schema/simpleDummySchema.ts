import {Rule} from '@sanity/types'
import {WorkshopSchemaProps, wrapSchema} from '../data'

export default function getSchema(props: WorkshopSchemaProps) {
  const {hiddenGroup = false} = props

  return wrapSchema({
    name: 'dummy',
    type: 'document',
    groups: [
      {
        name: 'group1',
        title: 'Group 1',
        hidden: hiddenGroup,
      },
    ],
    fields: [
      {
        name: 'title',
        title: 'Title',
        type: 'string',
      },
      {
        name: 'person',
        type: 'object',
        group: ['group1'],
        fieldsets: [
          {
            name: 'social',
            title: 'Social media handles [collapsed by default]',
            options: {collapsible: true, collapsed: true},
          },
        ],
        groups: [
          {
            name: 'instagram',
            title: 'Instagram',
          },
        ],
        fields: [
          {
            name: 'name',
            title: 'Name',
            type: 'string',
          },
          {
            name: 'conditionalName',
            title: 'Name (conditional read only)',
            type: 'string',
            readOnly: ({parent}: any) => parent?.name === 'ro',
          },
          {
            name: 'twitter',
            title: 'Twitter',
            type: 'string',
            fieldset: 'social',
            validation: (rule: Rule) => rule.required(),
          },
          {
            name: 'instagram',
            title: 'Instagram',
            type: 'string',
            fieldset: 'social',
            group: ['instagram'],
          },
          {
            name: 'facebook',
            title: 'Facebook',
            type: 'string',
            fieldset: 'social',
          },
        ],
      },
    ],
  })
}
