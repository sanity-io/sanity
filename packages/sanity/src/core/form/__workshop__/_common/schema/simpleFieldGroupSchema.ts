import {CogIcon} from '@sanity/icons'
import {WorkshopSchemaProps, wrapSchema} from '../data'

export default function getSchema(props: WorkshopSchemaProps) {
  const {hiddenGroup = false} = props

  return wrapSchema({
    name: 'dummy',
    title: 'Basic groups',
    type: 'document',
    groups: [
      {
        name: 'group1',
        title: 'Group 1',
        icon: CogIcon,
        hidden: hiddenGroup,
      },
      {
        name: 'group2',
        title: 'Group 2',
      },
      {
        name: 'group3',
        title:
          'Group 3 - with a very long name. How will it behave? It should not overflow the pane, but wrap.',
      },
    ],
    fields: [
      {
        name: 'title',
        title: 'Title',
        type: 'string',
      },
      {name: 'field1', type: 'string', group: 'group1'},
      {name: 'field2', type: 'string', group: 'group2'},
      {name: 'field3', type: 'string'},
      {name: 'field4', type: 'string', group: ['group1', 'group2']},
      {
        name: 'fieldGroup',
        type: 'object',
        group: 'group1',
        fields: [
          {
            name: 'groupField1',
            type: 'string',
          },
          {
            name: 'groupField2',
            type: 'string',
          },
        ],
      },
      {name: 'field5', type: 'string', group: 'group3'},
    ],
  })
}
