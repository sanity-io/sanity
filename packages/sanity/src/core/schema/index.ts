export * from './createSchema'
export * from './helpers'
