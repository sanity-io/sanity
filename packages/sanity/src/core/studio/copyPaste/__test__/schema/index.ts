import {type Schema} from '@sanity/types'

import {createSchema} from '../../../../schema'
import {authorDocument} from './author'
import {bookDocument} from './documents/book'
import {objectsDocument} from './documents/objects'
import {referencesDocument} from './documents/references'
import {editorDocument} from './editor'
import {hotspotDocument} from './hotspot'
import {linkType, myStringObjectType, nestedObjectType} from './objects'
import {postDocument} from './post'
import {pteCustomMarkersDocument} from './pteCustomerMarkers'

export const mockTypes = [
  linkType,
  myStringObjectType,
  nestedObjectType,
  {
    name: 'customNamedBlock',
    type: 'block',
    title: 'A named custom block',
    marks: {
      annotations: [linkType, myStringObjectType],
    },
    of: [
      {
        type: 'object',
        name: 'test',
        fields: [myStringObjectType],
      },
      {
        type: 'reference',
        name: 'strongAuthorRef',
        title: 'A strong author ref',
        to: {type: 'author'},
      },
    ],
  },
  authorDocument,
  editorDocument,
  postDocument,
  pteCustomMarkersDocument,
  hotspotDocument,
  objectsDocument,
  referencesDocument,
  bookDocument,
]

export const schema = createSchema({
  name: 'default',
  types: mockTypes,
}) as Schema
