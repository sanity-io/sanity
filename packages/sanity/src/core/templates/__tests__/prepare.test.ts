import {prepareTemplates, defaultTemplatesForSchema} from '../'
import {schema} from './schema'

describe('getTemplates', () => {
  test('returns defaults if part is not implemented', () => {
    const templates = defaultTemplatesForSchema(schema)
    expect(templates).toMatchSnapshot()
  })

  test('returns defined templates if part implemented', () => {
    const templates = prepareTemplates(schema, [
      {
        id: 'author',
        title: 'Author',
        schemaType: 'author',
        value: {title: 'here'},
      },
      {
        serialize: () => ({
          id: 'developer',
          title: 'Developer',
          schemaType: 'developer',
          value: {title: 'Foo'},
        }),
      },
    ] as any)

    expect(templates).toMatchSnapshot()
  })

  test('validates that templates has ID', () => {
    expect(() =>
      prepareTemplates(schema, [
        {
          title: 'Author',
          schemaType: 'author',
          value: {title: 'here'},
        },
      ] as any),
    ).toThrow('Template "Author" is missing required properties: id')
  })

  test('validates that templates has title', () => {
    expect(() =>
      prepareTemplates(schema, [
        {
          id: 'author',
          schemaType: 'author',
          value: {title: 'here'},
        },
      ] as any),
    ).toThrow('Template "author" is missing required properties: title')
  })

  test('validates that templates has schema type', () => {
    expect(() =>
      prepareTemplates(schema, [
        {
          id: 'author',
          title: 'Author',
          value: {title: 'here'},
        },
      ] as any),
    ).toThrow('Template "author" is missing required properties: schemaType')
  })

  test('validates that templates has value', () => {
    expect(() =>
      prepareTemplates(schema, [
        {
          id: 'author',
          title: 'Author',
          schemaType: 'author',
        },
      ] as any),
    ).toThrow('Template "author" is missing required properties: value')
  })

  test('validates that templates has id, title, schemaType, value', () => {
    expect(() => prepareTemplates(schema, [{}] as any)).toThrow(
      'Template at index 0 is missing required properties: id, title, schemaType, value',
    )
  })

  test('validates that templates has an object/function value', () => {
    expect(() =>
      prepareTemplates(schema, [
        {
          id: 'author',
          title: 'Author',
          schemaType: 'author',
          value: [],
        },
        {
          id: 'person',
          title: 'Person',
          schemaType: 'person',
          value: [],
        },
      ] as any),
    ).toThrow(
      'Template "author" has an invalid "value" property; should be a function or an object',
    )
  })

  test('validates that templates has unique IDs', () => {
    expect(() =>
      prepareTemplates(schema, [
        {
          id: 'author',
          title: 'Author',
          schemaType: 'author',
          value: {name: 'Gunnar'},
        },
        {
          id: 'person',
          title: 'Person',
          schemaType: 'person',
        },
      ] as any),
    ).toThrow('Template "person" is missing required properties: value')
  })
})
