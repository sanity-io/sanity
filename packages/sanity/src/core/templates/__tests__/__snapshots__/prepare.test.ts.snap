// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTemplates returns defaults if part is not implemented 1`] = `
Array [
  Object {
    "icon": [Function],
    "id": "author",
    "schemaType": "author",
    "title": "Author",
    "value": [Function],
  },
  Object {
    "icon": [Function],
    "id": "person",
    "schemaType": "person",
    "title": "Person",
    "value": Object {
      "_type": "person",
    },
  },
  Object {
    "icon": [Function],
    "id": "post",
    "schemaType": "post",
    "title": "Post",
    "value": Object {
      "_type": "post",
    },
  },
  Object {
    "icon": undefined,
    "id": "developer",
    "schemaType": "developer",
    "title": "Developer",
    "value": [Function],
  },
]
`;

exports[`getTemplates returns defined templates if part implemented 1`] = `
Array [
  Object {
    "id": "author",
    "schemaType": "author",
    "title": "Author",
    "value": Object {
      "title": "here",
    },
  },
  Object {
    "id": "developer",
    "schemaType": "developer",
    "title": "Developer",
    "value": Object {
      "title": "Foo",
    },
  },
]
`;
